internal/controllers/cluster/cluster_controller_status.go

```diff
@@ -0,0 +1,1099 @@
+		v1beta2conditions.NegativePolarityConditionTypes{clusterv1.ClusterScalingDownV1Beta2Condition},
+	)
+	if err != nil {
+		log.Error(err, "Failed to aggregate ControlPlane, MachinePool, MachineDeployment, MachineSet's ScalingDown conditions")
+		v1beta2conditions.Set(cluster, metav1.Condition{
+			Type:    clusterv1.ClusterScalingDownV1Beta2Condition,
+			Status:  metav1.ConditionUnknown,
+			Reason:  clusterv1.ClusterScalingDownInternalErrorV1Beta2Reason,
+			Message: "Please check controller logs for errors",
+		})
+		return
+	}
+
+	v1beta2conditions.Set(cluster, *workersAvailableCondition)
+}
+
+func setDeletingCondition(_ context.Context, cluster *clusterv1.Cluster, deletingReason, deletingMessage string) {
+	if cluster.DeletionTimestamp.IsZero() {
+		v1beta2conditions.Set(cluster, metav1.Condition{
+			Type:   clusterv1.ClusterDeletingV1Beta2Condition,
+			Status: metav1.ConditionFalse,
+			Reason: clusterv1.ClusterDeletingDeletionTimestampNotSetV1Beta2Reason,
+		})
+		return
+	}
+
+	v1beta2conditions.Set(cluster, metav1.Condition{
+		Type:    clusterv1.ClusterDeletingV1Beta2Condition,
+		Status:  metav1.ConditionTrue,
+		Reason:  deletingReason,
+		Message: deletingMessage,
+	})
+}
+
+type clusterConditionCustomMergeStrategy struct {
+	cluster                        *clusterv1.Cluster
+	negativePolarityConditionTypes sets.Set[string]
+}
+
+func (c clusterConditionCustomMergeStrategy) Merge(conditions []v1beta2conditions.ConditionWithOwnerInfo, conditionTypes []string) (status metav1.ConditionStatus, reason, message string, err error) {
+	return v1beta2conditions.DefaultMergeStrategyWithCustomPriority(func(condition metav1.Condition) v1beta2conditions.MergePriority {
+		// While cluster is deleting, treat unknown conditions from external objects as info (it is ok that those objects have been deleted at this stage).
+		if !c.cluster.DeletionTimestamp.IsZero() {
+			if condition.Type == clusterv1.ClusterInfrastructureReadyV1Beta2Condition && condition.Status == metav1.ConditionUnknown && (condition.Reason == clusterv1.ClusterInfrastructureDeletedV1Beta2Reason || condition.Reason == clusterv1.ClusterInfrastructureDoesNotExistV1Beta2Reason) {
+				return v1beta2conditions.InfoMergePriority
+			}
+			if condition.Type == clusterv1.ClusterControlPlaneAvailableV1Beta2Condition && condition.Status == metav1.ConditionUnknown && (condition.Reason == clusterv1.ClusterControlPlaneDeletedV1Beta2Reason || condition.Reason == clusterv1.ClusterControlPlaneDoesNotExistV1Beta2Reason) {
+				return v1beta2conditions.InfoMergePriority
+			}
+		}
+		return v1beta2conditions.GetDefaultMergePriorityFunc(c.negativePolarityConditionTypes)(condition)
+	}).Merge(conditions, conditionTypes)
+}
+
+func setAvailableCondition(ctx context.Context, cluster *clusterv1.Cluster) {
+	log := ctrl.LoggerFrom(ctx)
+
+	forConditionTypes := v1beta2conditions.ForConditionTypes{
+		clusterv1.ClusterInfrastructureReadyV1Beta2Condition,
```

Comment:
What about Deleting & ControlPlaneInitialized?

For Deleting, we would then also probably need an overrideCondition

Reply:
ControlPlaneInitialized should be implicitly included in ControlPlaneAvailable, but no objection to add it if we want.

For deleting, I think that what we have is acceptable (messages are way shorted than what we had in Machines); we can also defer to the refinement phase.

Reply:
Sounds good. If ControlPlaneAvailable covers the time when the CP is not initialized well, let's not add it (would just make aggregation harder)